src/string_object.cpp

```diff
@@ -1959,6 +1959,62 @@ Value StringObject::chop_in_place(Env *env) {
     return this;
 }
 
+Value StringObject::partition(Env *env, Value val) {
+    auto ary = new ArrayObject;
+
+    if (val->is_regexp()) {
+        auto match_result = val->as_regexp()->match(env, this);
+    
+        ssize_t start_byte_index;
+        ssize_t end_byte_index;
+
+        if (match_result->is_nil()){
+            auto empty = new StringObject { "" };
+            return new ArrayObject { this, empty, empty };
```

Comment:
Unfortunately we cannot use the same `empty` object twice:

```ruby
irb(main):002:0> ''.partition(/foo/).map(&:object_id)
=> [180, 200, 220]
```

Reply:
Actually, it looks like we cannot return `this` either! :cry: 

```ruby
irb(main):006:0> s.object_id
=> 240
irb(main):007:0> s.partition(/foo/).map(&:object_id)
=> [260, 280, 300]
```

Reply:
Hmm.. so what is the good solution?

Reply:
You'll just have to return an array with three new string objects, unfortunately:

```c++
return new ArrayObject { new StringObject, new StringObject, new StringObject };
```